### PR TITLE
feat(process): add process.{stdin, stdout, stderr} support

### PR DESCRIPTION
**Pull request by ThatOneBro**
Still trying to get pipe support and there seems to be some bugs in some edge cases. But overall main functionality is supported. Tests coming too.

### CODE REVIEW

File: src/bun.js/streams.exports.js [hunk 1]
```diff
@@ -3559,6 +3560,680 @@ var require_readable = __commonJS({
...
+          }
+          finishMaybe(this, state);
+        });
+      }
+    };
+    module.exports = WritableReadable;
+
+    function nop() {}
+    var kOnFinished = Symbol2("kOnFinished");
+    function WritableState(options, stream, isDuplex) {
```
Comment (Jarred-Sumner) [inline src/bun.js/streams.exports.js:3632]: This needs to be a class or it won't work properly

File: src/bun.js/streams.exports.js [hunk 2]
```diff
@@ -3559,6 +3560,680 @@ var require_readable = __commonJS({
...
+    WritableState.prototype.getBuffer = function getBuffer() {
+      return ArrayPrototypeSlice(this.buffered, this.bufferedIndex);
+    };
+    ObjectDefineProperty(WritableState.prototype, "bufferedRequestCount", {
+      get() {
+        return this.buffered.length - this.bufferedIndex;
+      },
+    });
+
+    ObjectDefineProperty(WritableReadable, SymbolHasInstance, {
```
Comment (Jarred-Sumner) [inline src/bun.js/streams.exports.js]: Define this on the class

File: src/bun.js/streams.exports.js [hunk 3]
```diff
@@ -3559,6 +3560,680 @@ var require_readable = __commonJS({
...
+      value: function (object) {
+        if (FunctionPrototypeSymbolHasInstance(this, object)) return true;
+        if (this !== Writable) return false;
+        return object && object._writableState instanceof WritableState;
+      },
+    });
+    WritableReadable.prototype.pipe = function () {
```
Comment (Jarred-Sumner) [inline src/bun.js/streams.exports.js]: Define this on the class

File: src/bun.js/streams.exports.js [hunk 4]
```diff
@@ -3559,6 +3560,680 @@ var require_readable = __commonJS({
...
+        } else if (i > 256) {
+          buffered.splice(0, i);
+          state.bufferedIndex = 0;
+        } else {
+          state.bufferedIndex = i;
+        }
+      }
+      state.bufferProcessing = false;
+    }
+    WritableReadable.prototype._write = function (chunk, encoding, cb) {
```
Comment (Jarred-Sumner) [inline src/bun.js/streams.exports.js]: Define this on the class

File: src/bun.js/streams.exports.js [hunk 5]
```diff
@@ -3559,6 +3560,680 @@ var require_readable = __commonJS({
...
+        const autoDestroy =
+          !rState ||
+          (rState.autoDestroy &&
+            (rState.endEmitted || rState.readable === false));
+        if (autoDestroy) {
+          stream.destroy();
+        }
+      }
+    }
+    ObjectDefineProperties(WritableReadable.prototype, {
```
Comment (Jarred-Sumner) [inline src/bun.js/streams.exports.js]: These can be set on the class too

File: src/bun.js/streams.exports.js [hunk 6]
```diff
@@ -4674,46 +5347,49 @@ var require_duplex = __commonJS({
     module.exports = Duplex;
 
     {
-      const keys = ObjectKeys(Writable.prototype);
+      const keys = ObjectKeys(WritableReadable.prototype);
       for (let i = 0; i < keys.length; i++) {
         const method = keys[i];
         if (!Duplex.prototype[method])
```
Comment (Jarred-Sumner) [inline src/bun.js/streams.exports.js:5525]: This calls the getter - we don't want to do that

Instead, let's do `method in prototype`

File: src/bun.js/streams.exports.js [hunk 7]
```diff
@@ -3559,6 +3560,680 @@ var require_readable = __commonJS({
...
+    var { errorOrDestroy } = destroyImpl;
+    var Readable = require_readable();
+
+    var WritableReadable = class WritableReadable extends Readable {
+      constructor(options = {}) {
+        super(options);
+
+        const isDuplex = this instanceof require_duplex();
+        if (!isDuplex && !FunctionPrototypeSymbolHasInstance(Writable, this))
+          return new Writable(options);
```
Comment (Jarred-Sumner) [inline src/bun.js/streams.exports.js]: This won't work

File: src/bun.js/streams.exports.js [hunk 8]
```diff
@@ -3559,6 +3560,680 @@ var require_readable = __commonJS({
...
+        this._writableState = new WritableState(options, this, isDuplex);
+        if (options) {
+          if (typeof options.write === "function") this._write = options.write;
```
Comment (Jarred-Sumner) [inline src/bun.js/streams.exports.js]: let's destructure `options` to minimize calls to getters

**Comment by Jarred-Sumner** [comment]
can you write tests for this that spawn bun in a separate process and have a permutation of `stdin`, `stderr`, and `stdout` which are:
- `pipe` (so the parent process reads it) with more than 65 KB of data pushed out / in
- `pipe` (so the parent process reads it) with less than 65 KB of data pushed out / in
- a file (to check it works when writing/reading files, like `bun ./foo > /tmp/123`)
- disabled (check that the process eventually closes)

**Comment by Jarred-Sumner** [comment]
the reason why 65 KB is relevant

that's the max capacity of a pipe

we want to test that the process doesn't block when we write more than the max capacity of a pipe

File: src/bun.js/process-stdio-polyfill.js [hunk 9]
```diff
@@ -0,0 +1,311 @@
...
+
+      #readable = true;
+      #writable = true;
+
+      #onFinish;
+      #onClose;
+      #onDrain;
+
+      fd = 0;
+      isTTY = true;
```
Comment (Jarred-Sumner) [inline src/bun.js/process-stdio-polyfill.js]: Missing `IsTTY`

```js
get isTTY() { 
   return require("tty").isatty(this.fd)
}
```

File: src/bun.js/process-stdio-polyfill.js [hunk 10]
```diff
@@ -0,0 +1,311 @@
...
+      #writeStream;
+
+      #readable = true;
+      #writable = true;
+
+      #onFinish;
+      #onClose;
+      #onDrain;
+
+      fd = 0;
```
Comment (Jarred-Sumner) [inline src/bun.js/process-stdio-polyfill.js:195]: `fd` should be read-only

In `constructor`

```js
super(options);
Object.defineProperty(this, "fd", {value: fd, writable: false, configurable: false});
```

File: src/bun.js/process-stdio-polyfill.js [hunk 11]
```diff
@@ -0,0 +1,311 @@
...
+      isTTY = true;
+
+      constructor() {
+        super({ readable: true, writable: true });
+
+        this.#readStream = Readable.fromWeb(Bun.stdin.stream());
```
Comment (Jarred-Sumner) [inline src/bun.js/process-stdio-polyfill.js]: Is this supposed to activate immediately? I thought it had to be explicitly `resume()` or have the handler attached

File: src/bun.js/process-stdio-polyfill.js [hunk 12]
```diff
@@ -0,0 +1,311 @@
...
+        this.#writeStream = createWriteStream("/dev/fd/0");
```
Comment (Jarred-Sumner) [inline src/bun.js/process-stdio-polyfill.js]: We probably need to rewrite `createWriteStream` to use the FileSink wrapper

File: src/bun.js/process-stdio-polyfill.js [hunk 13]
```diff
@@ -0,0 +1,311 @@
...
+
+        this.#writeStream.on("finish", () => {
```
Comment (Jarred-Sumner) [inline src/bun.js/process-stdio-polyfill.js]: Let's make these lazy

**Comment by Jarred-Sumner** [comment]
in what scenario is `stdin` writable? when it's a TCP socket?

**Comment by ThatOneBro** [comment]
> in what scenario is stdin writable? when it's a TCP socket?

I believe that in node it's always writable, I just tried and even when stdin is a tty I can write to it

Interestingly, almost all stdio in node (except for direct file read/write to fs I believe) is actually done via `net.Socket`, even in the tty case (in which it extends both a `net.Socket` and a `tty` obj too). net.Socket can be a UNIX domain socket or a TCP socket as far as I can tell

File: src/bun.js/process-stdio-polyfill.js (hunk 11 above)
Comment (ThatOneBro) [reply src/bun.js/process-stdio-polyfill.js]: Yeah we can have it activate on adding an event listener for readable or data 👍 I think 
```js
on(ev, listener) {
  super.on(ev, listener);
  if (ev === "data" || ev === "readable")
    this.#readStream = Readable.fromWeb(Bun.stdin.stream());
  // ...rest of setup
}
```
or something like this should work, or what do you think?

File: src/bun.js/child_process.exports.js [hunk 14]
```diff
@@ -1129,18 +1136,26 @@ export class ChildProcess extends EventEmitter {
...
     }
 
-    this.emit("exit", null, signal);
     this.#maybeClose();
 
-    // TODO: Make this actually ensure the process has exited before returning
-    // await this.#handle.exited()
-    // return this.#handle.killed;
+    // TODO: Figure out how to make this conform to the Node spec...
+    // The problem is that the handle does not report killed until the process exits
```
Comment (Jarred-Sumner) [inline src/bun.js/child_process.exports.js:1142]: This should be fixed now

File: src/bun.js/child_process.exports.js [hunk 15]
```diff
@@ -1280,26 +1295,64 @@ function abortChildProcess(child, killSignal) {
 
 class WrappedFileSink extends EventEmitter {
```
Comment (Jarred-Sumner) [inline src/bun.js/child_process.exports.js:1296]: Shouldn't this extend from a Writable so you can pipe to it?

File: src/bun.js/child_process.exports.js [hunk 16]
```diff
@@ -1280,26 +1295,64 @@ function abortChildProcess(child, killSignal) {
...
   }
 
   write(data) {
-    this.#fileSink.write(data);
-    this.#fileSink.flush(true);
+    var fileSink = this.#fileSink;
+    var result = fileSink.write(data);
+
+    var then = result?.then;
+    if (isPromise(result) && then && isCallable(then)) {
```
Comment (Jarred-Sumner) [inline src/bun.js/child_process.exports.js:1310]: the `&& then && isCallable(then)` checks shouldn't be necessary I think

File: src/bun.js/streams.exports.js [hunk 17]
```diff
@@ -3664,25 +3775,40 @@ var require_writable = __commonJS({
...
         super(options);
 
         const isDuplex = this instanceof require_duplex();
-        if (!isDuplex && !FunctionPrototypeSymbolHasInstance(Writable, this))
-          return new Writable(options);
+        if (
+          !isDuplex &&
+          !FunctionPrototypeSymbolHasInstance(WritableReadable, this)
+        )
+          return new WritableReadable(options);
```
Comment (Jarred-Sumner) [inline src/bun.js/streams.exports.js:3802]: I don't think constructors let you do this?

File: src/bun.js/streams.exports.js [hunk 18]
```diff
@@ -3892,7 +4158,7 @@ var require_writable = __commonJS({
       state.length -= state.writelen;
       state.writelen = 0;
       if (er) {
-        Error.captureStackTrace(er);
```
Comment (Jarred-Sumner) [inline src/bun.js/streams.exports.js:3895]: we will need to do Error.captureStackTrace for Bun

**Comment by Jarred-Sumner** [comment]
I don't feel like this is ready

I'm merging it anyway.

File: src/bun.js/child_process.exports.js (hunk 15 above)
Comment (ThatOneBro) [reply src/bun.js/child_process.exports.js:1296]: Good point, I think we can just use an fs.WriteStream now

**Pull request merged**